### PR TITLE
Challenge checker service

### PR DESCRIPTION
**Pull request by h4shk4t**

### CODE REVIEW

File: services/challengecheckerservice/deployChallCheck.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"fmt"
+	"time"
+
+	g "github.com/sdslabs/katana/configs"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func DeployChallChecker(chall_name, port, team_namespace string) error {
```
Comment (Bisht13) [inline services/challengecheckerservice/deployChallCheck.go:14]: Naming convention

File: services/challengecheckerservice/deployChallCheck.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	if err := GetClient(g.KatanaConfig.KubeConfig); err != nil {
+		return err
+	}
+	cronJobSpec := &batchv1.CronJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      chall_name + "-checker",
+			Namespace: namespace,
+		},
+		Spec: batchv1.CronJobSpec{
+			Schedule:          "*/1 * * * *", // Every minute.
```
Comment (Bisht13) [inline services/challengecheckerservice/deployChallCheck.go:26]: Make this a variable in config and get it from there

File: services/challengecheckerservice/deployChallCheck.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+							RestartPolicy: corev1.RestartPolicyOnFailure,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	cronJob, _ := kubeclient.BatchV1().CronJobs(namespace).Get(context.TODO(), chall_name+"-checker", metav1.GetOptions{})
+	if cronJob.Name == chall_name+"-checker" {
```
Comment (Bisht13) [inline services/challengecheckerservice/deployChallCheck.go:54]: What about updating challenge checker?

File: services/challengecheckerservice/deployChallCheck.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		fmt.Println("Challenge checker already exists for the challenge " + chall_name + " in namespace " + namespace)
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
```
Comment (Bisht13) [inline services/challengecheckerservice/deployChallCheck.go:59]: Timeout to be taken from config file

File: services/challengecheckerservice/deployChallCheck.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	defer cancel()
+
+	_, err := kubeclient.BatchV1().CronJobs(namespace).Create(ctx, cronJobSpec, metav1.CreateOptions{})
+	if err != nil {
+		panic(err.Error())
+	}
+
+	fmt.Println("Cron job created successfully!")
+
+	// if err != nil {
```
Comment (Bisht13) [inline services/challengecheckerservice/deployChallCheck.go:69]: Remove unnecessary comments

File: services/challengecheckerservice/helper.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"os"
+	"path/filepath"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var kubeclient *kubernetes.Clientset
```
Comment (Bisht13) [inline services/challengecheckerservice/helper.go:11]: Not needed, use from utils

File: services/challengecheckerservice/helper.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+// Get kubernetes client
+func GetClient(pathToCfg string) error {
```
Comment (Bisht13) [inline services/challengecheckerservice/helper.go:14]: Remove

File: services/master/controllers/deployCheckers.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
+package controllers
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/mholt/archiver/v3"
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:8]: Remove the unnecessary imports

File: services/master/controllers/deployCheckers.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	checker "github.com/sdslabs/katana/services/challengecheckerservice"
+)
+
+func DeployCheckers(c *fiber.Ctx) error {
+	foldername := ""
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:13]: Naming Convention

File: services/master/controllers/deployCheckers.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/mholt/archiver/v3"
+	checker "github.com/sdslabs/katana/services/challengecheckerservice"
+)
+
+func DeployCheckers(c *fiber.Ctx) error {
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:12]: should it be called DeployCheckers? or DeployChecker

File: services/master/controllers/deployCheckers.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	foldername := ""
+	fmt.Println("Starting")
+	if form, err := c.MultipartForm(); err == nil {
+
+		//sort this
+		if token := form.Value["token"]; len(token) > 0 {
+			// Get key value:
+			fmt.Println(token[0])
+			c.SendString("Test a")
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:21]: This code feels incomplete, finish it

File: services/master/controllers/deployCheckers.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+				// return c.SendString(url)
+				return c.SendString("Checker deployed successfully")
+			}
+		}
+	}
+	fmt.Println("Ending")
+	return c.SendString("Wrong file")
+}
+
+// func DeleteChallenge(c *fiber.Ctx) error {
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:114]: Remove all the commented code

**Review by Bisht13 — CHANGES_REQUESTED** [review]

File: services/master/controllers/deployCheckers.go (hunk 11 above)
Comment (h4shk4t) [reply services/master/controllers/deployCheckers.go:21]: I referred Ashpect's branch here... I will see the changes he made here and apply them.

File: services/master/controllers/deployCheckers.go (hunk 10 above)
Comment (h4shk4t) [reply services/master/controllers/deployCheckers.go:12]: Multiple checkers can be deployed at once, so I kept it DeployCheckers

Comment (Bisht13) [reply services/master/controllers/deployCheckers.go:12]: I get it but in totally there would be one checker for a challenge right?

File: services/master/controllers/deployCheckers.go (hunk 11 above)
Comment (Bisht13) [reply services/master/controllers/deployCheckers.go:21]: Rebase first, I've merged Percy's PR

**Review by Bisht13 — CHANGES_REQUESTED** [review]

File: services/master/controllers/gitServer.go [hunk 13]
```diff
@@ -77,6 +77,6 @@ func GitServer(c *fiber.Ctx) error {
 	if resp.StatusCode != http.StatusOK {
 		return fmt.Errorf("error while setting up Git Server")
 	}
-
+
```
Comment (Bisht13) [inline services/master/controllers/gitServer.go:80]: Redundant change

File: services/master/controllers/deployCheckers.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/gofiber/fiber/v2"
+	archiver "github.com/mholt/archiver/v3"
+	checker "github.com/sdslabs/katana/services/challengecheckerservice"
+)
+
+func deleteFolder(folderName string) (message string) {
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:13]: This should be in utils with a more generic implementation

File: services/master/controllers/deployCheckers.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+		return "Error deleting challenge folder"
+	}
+	fmt.Println("Challenge folder deleted")
+	return "Challenge folder deleted"
+}
+
+func DeployChecker(c *fiber.Ctx) error {
+	folderName := ""
+	fmt.Println("Starting")
+	if form, err := c.MultipartForm(); err == nil {
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:37]: Why is it a MultipartForm?

File: services/master/controllers/deployCheckers.go [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+		checkerFiles := form.File["challengeChecker"]
+
+		// Loops through all challenges, if multiple uploaded :
+		for _, file := range checkerFiles {
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:41]: Why would there be multiple challenges at once?

File: services/master/controllers/deployCheckers.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+			pattern := `([^/]+)\.tar\.gz$`
+			regex := regexp.MustCompile(pattern)
+			match := regex.FindStringSubmatch(file.Filename)
+			folderName = match[1]
+
+			response, _ := createfolder(folderName + "-checker")
+			if response == 1 {
+				fmt.Println("Directory for challenge checker already exists with same name")
+				return c.SendString("Directory for challenge checker already exists with same name")
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:50]: Shorten the return message

File: services/master/controllers/deployCheckers.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+			} else if response == 2 {
+				fmt.Println("Issue with creating challenge checkor directory.Check permissions")
+				return c.SendString("Issue with creating challenge checkor directory.Check permissions")
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:53]: Shorten the return message and I don't think so permissions would have to do anything with this?

File: services/master/controllers/deployCheckers.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+			}
+
+			if err := c.SaveFile(file, fmt.Sprintf("./challenges/%s/%s", folderName+"-checker", file.Filename)); err != nil {
+				return err
+			}
+
+			err = archiver.Unarchive("./challenges/"+folderName+"-checker/"+file.Filename, "./chall/"+folderName+"-checker")
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:60]: `file.Filename, "./chall/"+folderName+"-checker"` why?

File: services/master/controllers/deployCheckers.go [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+			if err != nil {
+				fmt.Println("Error in unarchiving", err)
+				return c.SendString("Error in unarchiving")
+			}
+
+			fmt.Println("Building docker image with tag", folderName+"-checker")
+			buildimage(folderName + "-checker")
+			fmt.Println("Docker images built successfully")
+
+			checker.DeployChallChecker(folderName, "5000", "katana-team-0-ns")
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:70]: Why is this hardcoded?

File: services/master/controllers/deployCheckers.go [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+			if err != nil {
+				return c.SendString(err.Error())
+			} else {
+				// return c.SendString(url)
+				return c.SendString("Checker deployed successfully")
+			}
+		}
+	}
+	fmt.Println("Ending")
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:79]: Remove

File: services/master/controllers/deployCheckers.go [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+	return c.SendString("Wrong file")
+}
+
+func UpdateChecker(c *fiber.Ctx) error {
+	folderName := ""
+	if form, err := c.MultipartForm(); err == nil {
```
Comment (Bisht13) [inline services/master/controllers/deployCheckers.go:85]: So much of common code

File: services/master/controllers/deploy.go [hunk 23]
```diff
@@ -20,6 +20,7 @@ func challcopy(dirPath, challengename, challengetype string) {
 	pathInPod := "/opt/katana/katana_" + challengetype + "_" + challengename + ".tar.gz"
 	fmt.Println("Testing" + localFilePath + "....and..." + pathInPod)
 	deployer.CopyInPod(localFilePath, pathInPod)
+	deployer.CopyInPod(localFilePath, pathInPod)
```
Comment (Bisht13) [inline services/master/controllers/deploy.go:23]: Is this needed?

File: services/challengecheckerservice/deployChallCheck.go [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+	"time"
+
+	g "github.com/sdslabs/katana/configs"
+	utils "github.com/sdslabs/katana/lib/utils"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func DeployChallChecker(challName, port, teamNamespace string) error {
```
Comment (Bisht13) [inline services/challengecheckerservice/deployChallCheck.go:15]: Change name to DeployChallengeChecker, never use chall for challenge

File: services/challengecheckerservice/deployChallCheck.go [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(g.KatanaConfig.TimeOut)*time.Second)
+	defer cancel()
+
+	_, err := kubeclient.BatchV1().CronJobs(namespace).Create(ctx, cronJobSpec, metav1.CreateOptions{})
+	if err != nil {
+		panic(err.Error())
+	}
+
+	fmt.Println("Cron job created successfully!")
```
Comment (Bisht13) [inline services/challengecheckerservice/deployChallCheck.go:64]: CronJob is one word

File: services/challengecheckerservice/deleteChecker.go [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+	"log"
+
+	g "github.com/sdslabs/katana/configs"
+	utils "github.com/sdslabs/katana/lib/utils"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func DeleteChecker(checkerName string) error {
+	namespace := "katana"
+	fmt.Println("Creating kube client...")
```
Comment (Bisht13) [inline services/challengecheckerservice/deleteChecker.go:15]: Remove

File: services/challengecheckerservice/deleteChecker.go [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+	kubeclient, err := utils.GetKubeClient(g.KatanaConfig.KubeConfig)
+	if err != nil {
+		log.Fatal("failed to create kube client", err)
```
Comment (Bisht13) [inline services/challengecheckerservice/deleteChecker.go:18]: you are not creating kube client, you are fetching it

File: services/challengecheckerservice/deleteChecker.go [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+	}
+	fmt.Println("Kube client created...")
```
Comment (Bisht13) [inline services/challengecheckerservice/deleteChecker.go:20]: Remove

File: services/challengecheckerservice/deleteChecker.go [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+	fmt.Println("---------------Deleting checker for challenge: ", checkerName)
+	cronJobClient := kubeclient.BatchV1().CronJobs(namespace)
+	propagationPolicy := metav1.DeletePropagationBackground
+	err = cronJobClient.Delete(context.Background(), checkerName, metav1.DeleteOptions{PropagationPolicy: &propagationPolicy})
+
+	if err != nil {
+
```
Comment (Bisht13) [inline services/challengecheckerservice/deleteChecker.go:27]: Unnecessary spaces

File: services/challengecheckerservice/deleteChecker.go [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+		log.Fatal("failed to delete cron job", err)
+
```
Comment (Bisht13) [inline services/challengecheckerservice/deleteChecker.go:29]: Delete line

File: services/challengecheckerservice/deleteChecker.go [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+	}
+	fmt.Println("Kube client created...")
+	fmt.Println("---------------Deleting checker for challenge: ", checkerName)
+	cronJobClient := kubeclient.BatchV1().CronJobs(namespace)
+	propagationPolicy := metav1.DeletePropagationBackground
+	err = cronJobClient.Delete(context.Background(), checkerName, metav1.DeleteOptions{PropagationPolicy: &propagationPolicy})
+
+	if err != nil {
+
+		log.Fatal("failed to delete cron job", err)
```
Comment (Bisht13) [inline services/challengecheckerservice/deleteChecker.go:28]: CronJob is one word

**Review by Bisht13 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**